include/emulator/cpu.h

```diff
@@ -1,162 +1,423 @@
-    /* 0x120CC */ char unk_0x120CC[0x150];
-    /* 0x1221C */ s32 unk_0x1221C;
-    /* 0x12220 */ s32 unk_0x12220;
-    /* 0x12224 */ s32 unk_0x12224;
-    /* 0x12228 */ s32 unk_0x12228;
-    /* 0x1222C */ s32 unk_0x1222C;
-    /* 0x12230 */ s32 unk_0x12230;
-    /* 0x12234 */ s32 unk_0x12234;
-    /* 0x12238 */ s32 unk_0x12238;
-    /* 0x1223C */ s32 unk_0x1223C;
-    /* 0x12240 */ s32 unk_0x12240;
-    /* 0x12244 */ s32 unk_0x12244;
-    /* 0x12248 */ s32 unk_0x12248;
-    /* 0x1224C */ s32 unk_0x1224C;
-    /* 0x12250 */ s32 unk_0x12250;
-    /* 0x12254 */ s32 unk_0x12254;
-    /* 0x12258 */ s32 unk_0x12258;
-    /* 0x1225C */ s32 unk_0x1225C;
-    /* 0x12260 */ s32 unk_0x12260;
-    /* 0x12264 */ s32 unk_0x12264;
-    /* 0x12268 */ s32 unk_0x12268;
-    /* 0x1226C */ s32 unk_0x1226C;
-    /* 0x12270 */ s32 unk_0x12270;
-    /* 0x12274 */ s32 unk_0x12274;
-    /* 0x12278 */ s32 tmp_mips_dest_reg;
-    /* 0x1227C */ s32 tmp_ppc_dest_reg;
-    /* 0x12280 */ s32 unk_0x12280;
-    /* 0x12284 */ s32 unk_0x12284;
-    /* 0x12288 */ s32 unk_0x12288;
-    /* 0x1228C */ s32 unk_0x1228C;
-    /* 0x12290 */ u32 unk_0x12290;
-    /* 0x12294 */ s32 unk_0x12294;
-    /* 0x12298 */ char unk_0x12298[0x38];
-} cpu_class_t; // size = 0x122D8
-
-s32 cpuSetGetBlock(cpu_class_t* cpu, cpu_dev_t* dev, void* arg2);
-s32 cpuSetDevicePut(cpu_class_t* cpu, cpu_dev_t* dev, sb_func_t sb, sh_func_t sh, sw_func_t sw, sd_func_t sd);
-s32 cpuSetDeviceGet(cpu_class_t* cpu, cpu_dev_t* dev, lb_func_t lb, lh_func_t lh, lw_func_t lw, ld_func_t ld);
-s32 cpuGetAddressBuffer(cpu_class_t *cpu, void **buffer, u32 addr);
+#include "emulator/xlObject.h"
+#include "revolution/os.h"
+#include "revolution/types.h"
+#include "macros.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// MIPS instruction encoding:
+// R-type: opcode (6 bits) | rs (5 bits) | rt (5 bits) | rd (5 bits) | sa (5 bits) | funct (6 bits)
+// I-type: opcode (6 bits) | rs (5 bits) | rt (5 bits) | imm (16 bits)
+// J-type: opcode (6 bits) | target (26 bits)
+//  float: opcode (6 bits) | fmt (5 bits) | ft (5 bits) | fs (5 bits) | fd (5 bits) | funct (6 bits)
+#define MIPS_OP(inst) ((inst) >> 26)
+#define MIPS_RS(inst) (((inst) >> 21) & 0x1F)
+#define MIPS_RT(inst) (((inst) >> 16) & 0x1F)
+#define MIPS_RD(inst) (((inst) >> 11) & 0x1F)
+#define MIPS_SA(inst) (((inst) >> 6) & 0x1F)
+#define MIPS_FUNCT(inst) ((inst) & 0x3F)
+#define MIPS_IMM_S16(inst) ((s16)((inst) & 0xFFFF))
+#define MIPS_IMM_U16(inst) ((u16)((inst) & 0xFFFF))
+#define MIPS_TARGET(inst) ((inst) & 0x3FFFFFF)
+
+#define MIPS_FMT(inst) (((inst) >> 21) & 0x1F)
+#define MIPS_FT(inst) (((inst) >> 16) & 0x1F)
+#define MIPS_FS(inst) (((inst) >> 11) & 0x1F)
+#define MIPS_FD(inst) (((inst) >> 6) & 0x1F)
+
+#define TLB_PGSZ_MASK 0x01FFE000
+#define TLB_PGSZ_4K 0
+#define TLB_PGSZ_16K 0x6000
+#define TLB_PGSZ_64K 0x1E000
+#define TLB_PGSZ_256K 0x7E000
+#define TLB_PGSZ_1M 0x1FE000
+#define TLB_PGSZ_4M 0x7FE000
+#define TLB_PGSZ_16M 0x1FFE000
```

Comment:
I think mips.h might have been a better home for these too. Eventually we can sync it back into oot-gc